File: backend/app/models/analysis_filters.py
from datetime import date, timedelta
from typing import Literal

from pydantic import BaseModel, Field


ComparePeriod = Literal["previous_month", "previous_year", "previous_period"]


class AnalysisFilters(BaseModel):
    date_from: date | None = None
    date_to: date | None = None
    providers: list[str] = Field(default_factory=list)
    account_ids: list[str] = Field(default_factory=list)
    include_transfers: bool = False
    compare_account_a: str | None = None
    compare_account_b: str | None = None
    compare_period: ComparePeriod = "previous_month"

    def resolved_date_range(self) -> tuple[date, date]:
        today = date.today()
        end = self.date_to or today
        start = self.date_from or (end - timedelta(days=180))
        if start > end:
            start, end = end, start
        return start, end

    def to_dict(self) -> dict:
        start, end = self.resolved_date_range()
        return {
            "date_from": start.isoformat(),
            "date_to": end.isoformat(),
            "providers": self.providers,
            "account_ids": self.account_ids,
            "include_transfers": self.include_transfers,
            "compare_account_a": self.compare_account_a,
            "compare_account_b": self.compare_account_b,
            "compare_period": self.compare_period,
        }


def parse_analysis_filters(
    *,
    date_from: str | None = None,
    date_to: str | None = None,
    provider: list[str] | None = None,
    account_id: list[str] | None = None,
    include_transfers: bool = False,
    compare_account_a: str | None = None,
    compare_account_b: str | None = None,
    compare_period: str = "previous_month",
) -> AnalysisFilters:
    cp: ComparePeriod = "previous_month"
    if compare_period in ("previous_month", "previous_year", "previous_period"):
        cp = compare_period  # type: ignore[assignment]

    return AnalysisFilters(
        date_from=date.fromisoformat(date_from) if date_from else None,
        date_to=date.fromisoformat(date_to) if date_to else None,
        providers=[p.lower() for p in (provider or []) if p],
        account_ids=list(account_id or []),
        include_transfers=include_transfers,
        compare_account_a=compare_account_a,
        compare_account_b=compare_account_b,
        compare_period=cp,
    )

File: frontend/lib/analysis-filters.ts
import { bankSourceFilterLabel } from "@/lib/provider-labels";

export type ComparePeriod = "previous_month" | "previous_year" | "previous_period";
export type ProviderFilter = "" | "plaid" | "mono";
export type DatePreset = "30d" | "90d" | "6m" | "ytd" | "custom";

export interface AnalysisFilterState {
  dateFrom: string;
  dateTo: string;
  provider: ProviderFilter;
  accountIds: string[];
  includeTransfers: boolean;
  comparePeriod: ComparePeriod;
  compareAccountA: string;
  compareAccountB: string;
}

export interface AnalyticsMetaAccount {
  id: string;
  account_name: string;
  provider: string;
  currency: string;
}

export const ANALYSIS_FILTERS_STORAGE_KEY = "finsight-analysis-filters";

export function getDefaultAnalysisFilters(): AnalysisFilterState {
  const today = new Date();
  const sixMonthsAgo = new Date(today);
  sixMonthsAgo.setMonth(sixMonthsAgo.getMonth() - 6);
  return {
    dateFrom: sixMonthsAgo.toISOString().slice(0, 10),
    dateTo: today.toISOString().slice(0, 10),
    provider: "",
    accountIds: [],
    includeTransfers: false,
    comparePeriod: "previous_month",
    compareAccountA: "",
    compareAccountB: "",
  };
}

export function loadAnalysisFilters(): AnalysisFilterState {
  const defaults = getDefaultAnalysisFilters();
  if (typeof window === "undefined") return defaults;
  try {
    const raw = localStorage.getItem(ANALYSIS_FILTERS_STORAGE_KEY);
    if (!raw) return defaults;
    return { ...defaults, ...JSON.parse(raw) };
  } catch {
    return defaults;
  }
}

export function saveAnalysisFilters(filters: AnalysisFilterState) {
  if (typeof window === "undefined") return;
  localStorage.setItem(ANALYSIS_FILTERS_STORAGE_KEY, JSON.stringify(filters));
}

export function buildAnalysisQueryString(filters: AnalysisFilterState): string {
  const params = new URLSearchParams();
  if (filters.dateFrom) params.set("date_from", filters.dateFrom);
  if (filters.dateTo) params.set("date_to", filters.dateTo);
  if (filters.provider) params.append("provider", filters.provider);
  filters.accountIds.forEach((id) => params.append("account_id", id));
  if (filters.includeTransfers) params.set("include_transfers", "true");
  if (filters.comparePeriod) params.set("compare_period", filters.comparePeriod);
  if (filters.compareAccountA) params.set("compare_account_a", filters.compareAccountA);
  if (filters.compareAccountB) params.set("compare_account_b", filters.compareAccountB);
  return params.toString();
}

export function applyDatePreset(preset: DatePreset): Pick<AnalysisFilterState, "dateFrom" | "dateTo"> {
  const today = new Date();
  const to = today.toISOString().slice(0, 10);
  const from = new Date(today);
  if (preset === "30d") from.setDate(from.getDate() - 30);
  else if (preset === "90d") from.setDate(from.getDate() - 90);
  else if (preset === "6m") from.setMonth(from.getMonth() - 6);
  else if (preset === "ytd") from.setMonth(0, 1);
  return { dateFrom: from.toISOString().slice(0, 10), dateTo: to };
}

export function filterSummaryChips(
  filters: AnalysisFilterState,
  accounts: AnalyticsMetaAccount[]
): string[] {
  const chips: string[] = [];
  if (filters.dateFrom && filters.dateTo) {
    chips.push(`${filters.dateFrom} → ${filters.dateTo}`);
  }
  if (filters.provider) {
    chips.push(bankSourceFilterLabel(filters.provider));
  }
  if (filters.accountIds.length > 0) {
    const names = filters.accountIds
      .map((id) => accounts.find((a) => a.id === id)?.account_name || id)
      .join(", ");
    chips.push(names);
  } else {
    chips.push("All accounts");
  }
  chips.push(filters.includeTransfers ? "incl. transfers" : "excl. transfers");
  const periodLabels: Record<ComparePeriod, string> = {
    previous_month: "vs last month",
    previous_year: "vs last year",
    previous_period: "vs prior period",
  };
  chips.push(periodLabels[filters.comparePeriod]);
  return chips;
}
